### PR TITLE
homed: flush fscrypt key on lock/deactivate

### PR DESCRIPTION
**Pull request by bluca**
The fscrypt key is added to the user keyring, and needs to be flushed out too.

Fixes https://github.com/systemd/systemd/issues/33138

### CODE REVIEW

File: src/home/homework-fscrypt.c [hunk 1]
```diff
@@ -29,6 +30,111 @@
...
+                                                ((key_serial_t *) keyring)[i]);
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        if ((size_t) c <= sz)
+                                break;
+
+                        sz = c;
+                        free(tuple);
+                }
```
Comment (poettering) [inline src/home/homework-fscrypt.c:70]: we have this same loop twice now, can we please get a helper call for this, i.e. add a helper to src/shared/keyring-util.c called keyring_describe() or so, that does this describe loop.

**Comment by poettering** [comment]
I am not sure i grok why this is even necessary. i.e. the per-user keyring should go away automatically when the kernel does not see any process owned by the keyring owner's key anymore.

hence, how come the keyring remains pinned for the original poster?

**Comment by poettering** [comment]
I don't think we need to hurry this into v256. This isn't really a regression.

**Comment by bluca** [comment]
> I am not sure i grok why this is even necessary. i.e. the per-user keyring should go away automatically when the kernel does not see any process owned by the keyring owner's key anymore.
> 
> hence, how come the keyring remains pinned for the original poster?

I don't know - but it's easily reproducible, it is there with the home deactivated and the user logged out

**Comment by bluca** [comment]
> I don't think we need to hurry this into v256. This isn't really a regression.

It isn't, however one of the key selling point of homed is to drop private key material from memory when the user is logged out, so I thought it would be good to fix anyway since it's not too difficult. It's also self-contained to homed+fscrypt and all done in a forked process, so should be as safe as it could be.

**Comment by poettering** [comment]
> I don't know - but it's easily reproducible, it is there with the home deactivated and the user logged out

well, what i am saying is that there probably is something else wrong here, if the the thing doesn#t get cleaned up automatically by the kernel

File: src/home/homework-fscrypt.c [hunk 2]
```diff
@@ -29,6 +30,111 @@
...
+                int c = -1; /* Workaround for maybe-uninitialized false positive due to missing_syscall indirection */
+
+                for (;;) {
+                        tuple = new(char, sz);
+                        if (!tuple) {
+                                log_oom();
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        c = keyctl(KEYCTL_DESCRIBE, ((key_serial_t *) keyring)[i], (unsigned long) tuple, c, 0);
```
Comment (poettering) [inline src/home/homework-fscrypt.c]: i think we should handle ENOKEY gracefully here (i.e. just silently skip over it, after all we might race against something else that deletes keys)

File: src/home/homework-fscrypt.c [hunk 3]
```diff
@@ -29,6 +30,111 @@
...
+                        if (c < 0) {
+                                log_debug_errno(r,
```
Comment (poettering) [inline src/home/homework-fscrypt.c]: first arg should be errno

File: src/home/homework-fscrypt.c [hunk 4]
```diff
@@ -29,6 +30,111 @@
...
+                                  tuple);
+                        _exit(EXIT_FAILURE);
+                }
+
+                if (!startswith(d + 1, "fscrypt:"))
+                        continue;
+
+                r = keyctl(KEYCTL_UNLINK, ((key_serial_t *) keyring)[i], KEY_SPEC_USER_KEYRING, 0, 0);
+                if (r < 0) {
+                        log_debug_errno(r,
```
Comment (poettering) [inline src/home/homework-fscrypt.c]: errno

**Comment by poettering** [comment]
So I am pretty sure there's some other bug lurking somewhere if the kernel doesn't clean this up automatically, but it might be ok to clean-up a user's keyring explicitly if we assume the user is fully logged out.

But I am pretty sure this should be dumbed down, and not be placed in homed, but logind, i.e. in user_finalize() at the same place where we call clean_ipc_by_uid().

This should really delete *all* keys of the user, i.e. fire a "clear" operation (KEYCTL_CLEAR).

**Comment by bluca** [comment]
> So I am pretty sure there's some other bug lurking somewhere if the kernel doesn't clean this up automatically, but it might be ok to clean-up a user's keyring explicitly if we assume the user is fully logged out.
> 
> But I am pretty sure this should be dumbed down, and not be placed in homed, but logind, i.e. in user_finalize() at the same place where we call clean_ipc_by_uid().
> 
> This should really delete _all_ keys of the user, i.e. fire a "clear" operation (KEYCTL_CLEAR).

Doing this in logind doesn't help with `homectl activate` -> `homectl deactivate` though, no? This specific issue is about the combination of homed+fscrypt, and it's homed that puts this key in the keyring, so it seems right to me that it's homed that removes it once it is not needed, just like it removes the other stuff from the keyrings. If issues with other things handled by logind show up then we can handle it there and then, but I think this fix should be targeted to this specific bug.

File: src/home/homework-fscrypt.c [hunk 5]
```diff
@@ -29,6 +30,98 @@
...
+int home_flush_keyring_fscrypt(UserRecord *h) {
+        _cleanup_free_ void *keyring = NULL;
+        size_t keyring_size = 0, n_keys = 0;
+        int r;
+
+        assert(user_record_storage(h) == USER_FSCRYPT);
+        assert(uid_is_valid(h->uid));
+
+        r = safe_fork("(sd-delkey)",
+                        FORK_RESET_SIGNALS|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT|FORK_REOPEN_LOG,
```
Comment (keszybz) [inline src/home/homework-fscrypt.c]: Indentation is off here.

File: src/home/homework-fscrypt.c [hunk 6]
```diff
@@ -29,6 +30,98 @@
...
+                        NULL);
+        if (r < 0)
+                return log_debug_errno(r,
+                                       "Failed delete encryption key from the keyring of user " UID_FMT ": %m",
+                                       h->uid);
```
Comment (keszybz) [inline src/home/homework-fscrypt.c]: Putting `r` on a separate line makes those logging calls very vertically spread out. We generally don't do that. It'd be nicer to make merge the first two lines.

The message should be "Failed to fork".

File: src/home/homework-fscrypt.c [hunk 7]
```diff
@@ -29,6 +30,98 @@
...
+        if (r > 0)
+                return 0;
+
+        /* Child */
+
+        r = fully_set_uid_gid(
```
Comment (keszybz) [inline src/home/homework-fscrypt.c:41]: It'd be nicer to move the implementation of child into a separate function that does a normal negative return. All those `_exit(EXIT_FAILURE)` calls are overly verbose and error-prone.

File: src/home/homework-fscrypt.c [hunk 8]
```diff
@@ -29,6 +30,97 @@
...
+                        h->uid,
+                        user_record_gid(h),
+                        /* supplementary_gids= */ NULL,
+                        /* n_supplementary_gids= */ 0);
+        if (r < 0)
+                return log_debug_errno(
+                                r,
+                                "Failed to change UID/GID to " UID_FMT "/" GID_FMT ": %m",
+                                h->uid,
+                                user_record_gid(h));
```
Comment (keszybz) [inline src/home/homework-fscrypt.c]: That's even worse now. What's wrong with
```C
return log_debug_errno(r, "Failed to change UID/GID to " UID_FMT "/" GID_FMT ": %m",
                       h->uid, user_record_gid(h));
```
?

File: src/home/homework-fscrypt.c [hunk 9]
```diff
@@ -29,6 +30,97 @@
...
+
+int home_flush_keyring_fscrypt(UserRecord *h) {
+        int r;
+
+        assert(user_record_storage(h) == USER_FSCRYPT);
+        assert(uid_is_valid(h->uid));
+
+        r = safe_fork("(sd-delkey)",
+                        FORK_RESET_SIGNALS|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT|FORK_REOPEN_LOG,
+                        NULL);
```
Comment (keszybz) [inline src/home/homework-fscrypt.c]: Indentation still off.

File: src/home/homework-fscrypt.c [hunk 10]
```diff
@@ -29,6 +30,97 @@
...
+                if (r < 0) {
+                        if (errno == ENOKEY) /* Something else deleted it already, that's ok. */
+                                continue;
+
+                        return log_debug_errno(
+                                        errno,
+                                        "Failed to delete encryption key with id '%d' from the keyring of user " UID_FMT
+                                        ": %m",
+                                        *key,
+                                        h->uid);
```
Comment (keszybz) [inline src/home/homework-fscrypt.c]: No, don't split message strings like this.

File: src/home/homework.c [hunk 11]
```diff
@@ -376,9 +376,16 @@ int keyring_unlink(key_serial_t k) {
 static int keyring_flush(UserRecord *h) {
         _cleanup_free_ char *name = NULL;
         long serial;
+        int r;
 
         assert(h);
 
+        if (user_record_storage(h) == USER_FSCRYPT && uid_is_valid(h->uid)) {
+                r = home_flush_keyring_fscrypt(h);
```
Comment (keszybz) [inline src/home/homework.c]: That function logs, so just `(void) home_flush_keyring_fscrypt(h)`.

**Review by keszybz — CHANGES_REQUESTED** [review]

**Comment by keszybz** [comment]
This looks better, but the (automatic, I assume) formatting is still rather ugly.

**Comment by bluca** [comment]
> This looks better, but the (automatic, I assume) formatting is still rather ugly.

Updated - but the issue is that this changes depending on who's reviewing. Some want multiple parameters per line starting on the same line, some want one parameter per line starting on the next line. We should really just get clang-formatter going and enforce it via CI and let that be the end of it...

**Comment by keszybz** [comment]
I think you're looking at this from the wrong angle. Code is something that is read much more than it is written. Thus, self-consistent, legible, clean code is worth a bit of extra formatting effort. Clang-format and other tools can be a useful tool, especially for beginners and a new codebase, but are not powerful enough for us.

FWIW, people tried to build a clang format specification before, and it always fell short.

**Comment by DaanDeMeyer** [comment]
> I think you're looking at this from the wrong angle. Code is something that is read much more than it is written. Thus, self-consistent, legible, clean code is worth a bit of extra formatting effort. Clang-format and other tools can be a useful tool, especially for beginners and a new codebase, but are not powerful enough for us.
> 
> FWIW, people tried to build a clang format specification before, and it always fell short.

I would much rather have us use clang-format as well. I couldn't care less that it isn't always perfect if only to save both longterm and new contributors the hassle of having to deal with formatting.

**Comment by bluca** [comment]
> > I think you're looking at this from the wrong angle. Code is something that is read much more than it is written. Thus, self-consistent, legible, clean code is worth a bit of extra formatting effort. Clang-format and other tools can be a useful tool, especially for beginners and a new codebase, but are not powerful enough for us.
> > FWIW, people tried to build a clang format specification before, and it always fell short.
> 
> I would much rather have us use clang-format as well. I couldn't care less that it isn't always perfect if only to save both longterm and new contributors the hassle of having to deal with formatting.

Exactly - the problem is that "self-consistent, legible, clean" are in the eye of the beholder, so depending on who happens to review a PR, requirements change, which is not great. If it's enforced by a tool, then you cannot argue against the tool (well you can try, but the tool will just look at you with a funny face)

**Comment by keszybz** [comment]
That'd mean having ugly formatting. If people want to use clang format or some other helper, that's fine. It's up to them, as long as they fix the most egregious shortcomings. But making this enforced is a completely different thing.  That'd force people who want and can do better to commit badly formatted code.

File: src/home/homework-fscrypt.c [hunk 12]
```diff
@@ -29,6 +30,93 @@
...
+        r = fully_set_uid_gid(
+                        h->uid,
+                        user_record_gid(h),
+                        /* supplementary_gids= */ NULL,
+                        /* n_supplementary_gids= */ 0);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to change UID/GID to " UID_FMT "/" GID_FMT ": %m",
+                                       h->uid, user_record_gid(h));
+
+        r = keyring_read(KEY_SPEC_USER_KEYRING, (void **) &keyring, &keyring_size);
```
Comment (poettering) [inline src/home/homework-fscrypt.c]: the cast seems unnecessary

File: src/home/homework-fscrypt.c [hunk 13]
```diff
@@ -29,6 +30,93 @@
...
+        if (r < 0)
+                return log_debug_errno(r, "Failed to read the keyring of user " UID_FMT ": %m", h->uid);
+
+        n_keys = keyring_size / sizeof(key_serial_t);
```
Comment (poettering) [inline src/home/homework-fscrypt.c:54]: maybe add: `assert(keyring_size % sizeof(key_serial_t) == 0)`

File: src/home/homework.c [hunk 14]
```diff
@@ -379,6 +379,9 @@ static int keyring_flush(UserRecord *h) {
 
         assert(h);
 
+        if (user_record_storage(h) == USER_FSCRYPT && uid_is_valid(h->uid))
```
Comment (poettering) [inline src/home/homework.c]: please move the uid_is_valid() check into home_flush_keyring_fscrypt()

File: src/home/homework-fscrypt.c [hunk 15]
```diff
@@ -29,6 +30,93 @@
 #include "user-util.h"
 #include "xattr-util.h"
 
+static int fscrypt_unlink_key(UserRecord *h) {
+        _cleanup_free_ void *keyring = NULL;
+        size_t keyring_size = 0, n_keys = 0;
+        int r;
+
+        assert(h);
```
Comment (poettering) [inline src/home/homework-fscrypt.c:38]: add: 
```c
assert(user_record_storage(h) == USER_FSCRYPT));
```

File: src/home/homework-fscrypt.c [hunk 16]
```diff
@@ -29,6 +30,93 @@
...
+        int r;
+
+        assert(user_record_storage(h) == USER_FSCRYPT);
+        assert(uid_is_valid(h->uid));
+
+        r = safe_fork("(sd-delkey)",
+                      FORK_RESET_SIGNALS|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT|FORK_REOPEN_LOG,
+                      NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to fork to delete fscrypt key from keyring: %m");
```
Comment (poettering) [inline src/home/homework-fscrypt.c]: you already set FORK_LOG, hence no need to log again, just drop this and do a simple `return r`

File: src/home/homework-fscrypt.c [hunk 17]
```diff
@@ -29,6 +30,93 @@
...
+        assert(h);
+
+        r = fully_set_uid_gid(
+                        h->uid,
+                        user_record_gid(h),
+                        /* supplementary_gids= */ NULL,
+                        /* n_supplementary_gids= */ 0);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to change UID/GID to " UID_FMT "/" GID_FMT ": %m",
+                                       h->uid, user_record_gid(h));
```
Comment (poettering) [inline src/home/homework-fscrypt.c:48]: should be at log_error_errno()

File: src/home/homework-fscrypt.c [hunk 18]
```diff
@@ -29,6 +30,93 @@
...
+
+        r = keyring_read(KEY_SPEC_USER_KEYRING, (void **) &keyring, &keyring_size);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to read the keyring of user " UID_FMT ": %m", h->uid);
```
Comment (poettering) [inline src/home/homework-fscrypt.c]: log_error_errno

File: src/home/homework-fscrypt.c [hunk 19]
```diff
@@ -29,6 +30,93 @@
...
+        FOREACH_ARRAY(key, ((key_serial_t *) keyring), n_keys) {
+                _cleanup_free_ char *description = NULL;
+                char *d;
+
+                r = keyring_describe(*key, &description);
+                if (r < 0) {
+                        if (r == -ENOKEY) /* Something else deleted it already, that's ok. */
+                                continue;
+
+                        return log_debug_errno(r, "Failed to describe key id %d: %m", *key);
```
Comment (poettering) [inline src/home/homework-fscrypt.c]: also LOG_ERR level

File: src/home/homework-fscrypt.c [hunk 20]
```diff
@@ -29,6 +30,93 @@
...
+                }
+
+                /* The decription is the final element as per manpage. */
+                d = strrchr(description, ';');
+                if (!d)
+                        return log_debug_errno(
+                                        SYNTHETIC_ERRNO(EINVAL),
+                                        "Failed to parse description of key id %d: %s",
+                                        *key,
+                                        description);
```
Comment (poettering) [inline src/home/homework-fscrypt.c:79]: ditto here (and everywhere else, i.e. hte functions in homework-fscrypt.c all log on their own)

File: src/home/homework-fscrypt.c [hunk 21]
```diff
@@ -29,6 +30,93 @@
...
+                                continue;
+
+                        return log_debug_errno(
+                                        errno,
+                                        "Failed to delete encryption key with id '%d' from the keyring of user " UID_FMT ": %m",
+                                        *key,
+                                        h->uid);
+                }
+
+                log_debug("Deleted encryption key with id '%d' from the keyring of user " UID_FMT ".", *key, h->uid);
```
Comment (poettering) [inline src/home/homework-fscrypt.c:96]: this one is ok to stay at debug level

**Comment by poettering** [comment]
lgtm

**Pull request merged**